File: src/quectel/parser.py
"""Parsers for Quectel AT command responses."""

from typing import Dict, Iterator, List, Optional, Tuple

from .models import (
    CarrierComponent,
    DeviceInfo,
    LteServingCell,
    NetworkInfo,
    NeighbourCell,
    Nr5gServingCell,
)


def parse_response(text: str, prefix: str) -> Iterator[List[str]]:
    """Parse AT command response lines with a common pattern.

    AT command responses follow a consistent format:
    - Empty lines to skip
    - Lines starting with prefix followed by colon, then comma-separated values
    - Values may be quoted

    Args:
        text: Raw response text from modem
        prefix: Command prefix to match (e.g., "+QSPN", "+QENG")

    Yields:
        List of stripped, unquoted values for each matching line
    """
    for line in text.splitlines():
        line = line.strip()
        if not line or line in ("OK", "ERROR"):
            continue

        if not line.startswith(prefix + ":"):
            continue

        # Remove prefix and colon
        content = line[len(prefix) + 1:].strip()

        # Split on comma and strip quotes/whitespace from each value
        values = []
        for val in content.split(","):
            val = val.strip()
            if val.startswith('"') and val.endswith('"'):
                val = val[1:-1]
            values.append(val)

        yield values


def parse_int(val: str) -> Optional[int]:
    """Parse integer, returning None for invalid/missing values."""
    if not val or val == "-":
        return None
    try:
        return int(val)
    except ValueError:
        return None


def parse_float(val: str) -> Optional[float]:
    """Parse float, returning None for invalid/missing values."""
    if not val or val == "-":
        return None
    try:
        return float(val)
    except ValueError:
        return None


def parse_hex(val: str) -> Optional[int]:
    """Parse hex string to int, returning None for invalid values."""
    if not val or val == "-":
        return None
    try:
        return int(val, 16)
    except ValueError:
        return None


def parse_ati(response: str) -> Optional[DeviceInfo]:
    """Parse ATI response.

    ATI is special - it doesn't follow the +PREFIX: format.
    It returns plain lines: manufacturer, model, revision.
    """
    lines = [l.strip() for l in response.splitlines() if l.strip()]
    lines = [l for l in lines if l not in ("OK", "ERROR")]

    if len(lines) < 3:
        return None

    manufacturer = lines[0]
    model = lines[1]
    revision = ""

    for line in lines[2:]:
        if line.startswith("Revision:"):
            revision = line.replace("Revision:", "").strip()
            break

    return DeviceInfo(
        manufacturer=manufacturer,
        model=model,
        revision=revision,
    )


def parse_qspn(response: str) -> Optional[NetworkInfo]:
    """Parse AT+QSPN response.

    Example: +QSPN: "I TIM","TIM","",0,"22201"
    """
    for values in parse_response(response, "+QSPN"):
        if len(values) < 5:
            continue

        full_name = values[0]
        short_name = values[1]
        mcc_mnc = values[4]

        if len(mcc_mnc) >= 5:
            mcc = int(mcc_mnc[:3])
            mnc = int(mcc_mnc[3:])
        else:
            mcc = 0
            mnc = 0

        return NetworkInfo(
            full_name=full_name,
            short_name=short_name,
            mcc=mcc,
            mnc=mnc,
        )

    return None


def parse_qeng_servingcell(
    response: str,
) -> Tuple[Optional[LteServingCell], Optional[Nr5gServingCell]]:
    """Parse AT+QENG="servingcell" response.

    Example lines:
        +QENG: "servingcell","NOCONN"
        +QENG: "LTE","FDD",222,01,328261F,280,275,1,4,4,BE3,-99,-14,-66,7,4,30,-
        +QENG: "NR5G-NSA",222,01,920,-96,18,-10,648768,78,10,1
    """
    lte_cell = None
    nr5g_cell = None

    for values in parse_response(response, "+QENG"):
        if len(values) < 2:
            continue

        # LTE serving cell
        if values[0] == "LTE" and len(values) >= 17:
            try:
                tx_power_raw = parse_float(values[16])
                tx_power = tx_power_raw / 10.0 if tx_power_raw is not None else None

                lte_cell = LteServingCell(
                    mode=values[1],
                    mcc=int(values[2]),
                    mnc=int(values[3]),
                    cell_id=parse_hex(values[4]) or 0,
                    pci=int(values[5]),
                    earfcn=int(values[6]),
                    band=int(values[7]),
                    ul_bandwidth_idx=parse_int(values[8]) or 0,
                    dl_bandwidth_idx=parse_int(values[9]) or 0,
                    tac=parse_hex(values[10]) or 0,
                    rsrp=parse_int(values[11]) or 0,
                    rsrq=parse_int(values[12]) or 0,
                    rssi=parse_int(values[13]) or 0,
                    sinr=parse_float(values[14]) or 0.0,
                    cqi=parse_int(values[15]),
                    tx_power=tx_power,
                )
            except (ValueError, IndexError):
                pass

        # NR5G-NSA serving cell
        elif "NR5G-NSA" in values[0] and len(values) >= 10:
            try:
                nr5g_cell = Nr5gServingCell(
                    mcc=int(values[1]),
                    mnc=int(values[2]),
                    pci=int(values[3]),
                    rsrp=int(values[4]),
                    sinr=float(values[5]),
                    rsrq=int(values[6]),
                    arfcn=int(values[7]),
                    band=int(values[8]),
                    bandwidth_idx=int(values[9]),
                )
            except (ValueError, IndexError):
                pass

    return (lte_cell, nr5g_cell)


def _pcell_state(code: str) -> str:
    """Convert PCell state code to string."""
    states = {
        "0": "Idle",
        "1": "Registered",
        "2": "Searching",
        "3": "Denied",
        "4": "Unknown",
        "5": "Roaming",
    }
    return states.get(code, "N/A")


def _scell_state(code: str) -> str:
    """Convert SCell state code to string."""
    states = {
        "0": "Deconfigured",
        "1": "Inactive",
        "2": "Active",
    }
    return states.get(code, "N/A")


def parse_qcainfo(response: str) -> List[CarrierComponent]:
    """Parse AT+QCAINFO response.

    Examples:
        +QCAINFO: "PCC",275,75,"LTE BAND 1",1,280,-99,-14,-67,-4
        +QCAINFO: "SCC",1350,100,"LTE BAND 3",1,240,-95,-18,-68,-10,0,-,-
        +QCAINFO: "SCC",648768,10,"NR5G BAND 78",920
    """
    components = []

    for values in parse_response(response, "+QCAINFO"):
        if len(values) < 5:
            continue

        c_type = values[0]
        earfcn = parse_int(values[1]) or 0
        bw_raw = parse_int(values[2]) or 0
        band_name = values[3]
        pci = 0
        state = None
        rsrp = None
        rsrq = None
        sinr = None
        ul_configured = None
        ul_band = None
        ul_earfcn = None

        try:
            if c_type == "PCC":
                # PCC,earfcn,bw,band,state,pci,rsrp,rsrq,rssi,sinr
                if len(values) >= 10:
                    state = _pcell_state(values[4])
                    pci = parse_int(values[5]) or 0
                    rsrp = parse_int(values[6])
                    rsrq = parse_int(values[7])
                    sinr = parse_float(values[9])
            else:  # SCC
                if len(values) == 5:
                    # NR5G SCC: SCC,arfcn,bw_idx,band,pci
                    pci = parse_int(values[4]) or 0
                elif len(values) == 9:
                    # SCC without signal: SCC,earfcn,bw,band,state,pci,ul_cfg,ul_band,ul_earfcn
                    state = _scell_state(values[4])
                    pci = parse_int(values[5]) or 0
                    ul_configured = parse_int(values[6])
                    ul_band = values[7] if values[7] != "-" else None
                    ul_earfcn = parse_int(values[8])
                elif len(values) == 12:
                    # SCC with signal but different order
                    state = _scell_state(values[4])
                    pci = parse_int(values[5]) or 0
                    ul_configured = parse_int(values[6])
                    ul_band = values[7] if values[7] != "-" else None
                    ul_earfcn = parse_int(values[8])
                    rsrp = parse_int(values[9])
                    rsrq = parse_int(values[10])
                    sinr = parse_float(values[11])
                elif len(values) >= 13:
                    # Full SCC
                    state = _scell_state(values[4])
                    pci = parse_int(values[5]) or 0
                    rsrp = parse_int(values[6])
                    rsrq = parse_int(values[7])
                    sinr = parse_float(values[9])
                    ul_configured = parse_int(values[10])
                    ul_band = values[11] if values[11] != "-" else None
                    ul_earfcn = parse_int(values[12])

            components.append(
                CarrierComponent(
                    component_type=c_type,
                    earfcn=earfcn,
                    bandwidth_raw=bw_raw,
                    band_name=band_name,
                    state=state,
                    pci=pci,
                    rsrp=rsrp,
                    rsrq=rsrq,
                    sinr=sinr,
                    ul_configured=ul_configured,
                    ul_band=ul_band,
                    ul_earfcn=ul_earfcn,
                )
            )
        except (ValueError, IndexError):
            continue

    return components


def parse_qeng_neighbourcell(response: str) -> List[NeighbourCell]:
    """Parse AT+QENG="neighbourcell" response.

    Examples:
        +QENG: "neighbourcell intra","LTE",275,280,-14,-99,-67,-,-,-,-,-,-
        +QENG: "neighbourcell inter","LTE",1350,240,-18,-95,-68,-,-,-,-,-
    """
    cells = []

    for values in parse_response(response, "+QENG"):
        if len(values) < 6:
            continue

        # First value is "neighbourcell intra" or "neighbourcell inter"
        cell_type_str = values[0]
        if "neighbourcell" not in cell_type_str:
            continue

        if "intra" in cell_type_str:
            cell_type = "intra"
        elif "inter" in cell_type_str:
            cell_type = "inter"
        else:
            continue

        try:
            tech = values[1]
            if tech != "LTE":
                continue

            cells.append(
                NeighbourCell(
                    cell_type=cell_type,
                    technology=tech,
                    earfcn=int(values[2]),
                    pci=int(values[3]),
                    rsrq=parse_int(values[4]) or 0,
                    rsrp=parse_int(values[5]) or 0,
                    rssi=parse_int(values[6]) or 0 if len(values) > 6 else 0,
                )
            )
        except (ValueError, IndexError):
            continue

    return cells


def parse_qnwprefcfg(response: str) -> Dict[str, str]:
    """Parse AT+QNWPREFCFG responses.

    Examples:
        +QNWPREFCFG: "mode_pref",AUTO
        +QNWPREFCFG: "nsa_nr5g_band",78
        +QNWPREFCFG: "lte_band",1:3:7:20
    """
    config = {}

    for values in parse_response(response, "+QNWPREFCFG"):
        if len(values) >= 2:
            key = values[0]
            value = values[1]
            config[key] = value

    return config

File: lua/quectel/parser.lua
-- AT command response parsing
-- Unified parsing approach for Quectel modem responses

local M = {}

--- Parse a single line of AT response, extracting values after prefix
-- Handles quoted strings and numeric values
-- @param line The response line
-- @param prefix The expected prefix (e.g., "+QENG")
-- @return Table of values, or nil if line doesn't match prefix
function M.parse_line(line, prefix)
    line = line:match("^%s*(.-)%s*$")  -- trim
    if line == "" or line == "OK" or line == "ERROR" then
        return nil
    end

    local expected = prefix .. ":"
    if line:sub(1, #expected) ~= expected then
        return nil
    end

    local content = line:sub(#expected + 1):match("^%s*(.-)%s*$")
    local values = {}

    -- Parse comma-separated values, handling quoted strings
    local pos = 1
    while pos <= #content do
        local char = content:sub(pos, pos)
        if char == '"' then
            -- Quoted string
            local end_quote = content:find('"', pos + 1)
            if end_quote then
                table.insert(values, content:sub(pos + 1, end_quote - 1))
                pos = end_quote + 1
                -- Skip comma
                if content:sub(pos, pos) == "," then
                    pos = pos + 1
                end
            else
                break
            end
        elseif char == "," then
            -- Empty value
            table.insert(values, "")
            pos = pos + 1
        else
            -- Unquoted value
            local comma = content:find(",", pos)
            local val
            if comma then
                val = content:sub(pos, comma - 1)
                pos = comma + 1
            else
                val = content:sub(pos)
                pos = #content + 1
            end
            val = val:match("^%s*(.-)%s*$")  -- trim
            table.insert(values, val)
        end
    end

    return values
end

--- Parse multi-line AT response, yielding all matching lines
-- @param text Full AT response text
-- @param prefix The expected prefix (e.g., "+QENG")
-- @return Iterator function yielding value tables
function M.parse_response(text, prefix)
    local lines = {}
    for line in text:gmatch("[^\r\n]+") do
        table.insert(lines, line)
    end

    local i = 0
    return function()
        while true do
            i = i + 1
            if i > #lines then return nil end
            local values = M.parse_line(lines[i], prefix)
            if values then return values end
        end
    end
end

--- Parse ATI response for device info
-- @param text ATI response
-- @return Table with manufacturer, model, revision
function M.parse_ati(text)
    local lines = {}
    for line in text:gmatch("[^\r\n]+") do
        line = line:match("^%s*(.-)%s*$")
        if line ~= "" and line ~= "OK" then
            table.insert(lines, line)
        end
    end

    return {
        manufacturer = lines[1] or "",
        model = lines[2] or "",
        revision = (lines[3] or ""):gsub("^Revision:%s*", "")
    }
end

--- Parse +QSPN response for operator info
-- @param text AT+QSPN response
-- @return Table with operator name, short name, mcc_mnc
function M.parse_qspn(text)
    for values in M.parse_response(text, "+QSPN") do
        return {
            operator = values[1] or "",
            operator_short = values[2] or "",
            mcc_mnc = values[5] or ""
        }
    end
    return nil
end

--- Parse +QENG="servingcell" response
-- @param text AT+QENG="servingcell" response
-- @return Table with serving cell info (LTE and optionally NR5G-NSA)
--
-- Handles two response formats:
--   Two-line:  +QENG: "servingcell","NOCONN"  /  +QENG: "LTE","FDD",...
--   One-line:  +QENG: "servingcell","NOCONN","LTE","FDD",...

-- Parse LTE fields from values array starting at offset o
-- o is the index of the duplex field (FDD/TDD)
local function parse_lte_fields(values, o)
    return {
        duplex = values[o],
        mcc = tonumber(values[o + 1]),
        mnc = tonumber(values[o + 2]),
        cell_id = values[o + 3],
        pci = tonumber(values[o + 4]),
        arfcn = tonumber(values[o + 5]),
        band = tonumber(values[o + 6]),
        bandwidth_dl = tonumber(values[o + 7]),
        bandwidth_ul = tonumber(values[o + 8]),
        tac = values[o + 9],
        rsrp = tonumber(values[o + 10]),
        rsrq = tonumber(values[o + 11]),
        rssi = tonumber(values[o + 12]),
        sinr = tonumber(values[o + 13]),
    }
end

-- Parse NR5G-NSA fields from values array starting at offset o
-- o is the index of the mcc field
local function parse_nr5g_fields(values, o)
    return {
        mcc = tonumber(values[o]),
        mnc = tonumber(values[o + 1]),
        pci = tonumber(values[o + 2]),
        rsrp = tonumber(values[o + 3]),
        sinr = tonumber(values[o + 4]),
        rsrq = tonumber(values[o + 5]),
        arfcn = tonumber(values[o + 6]),
        band = tonumber(values[o + 7]),
        bandwidth = tonumber(values[o + 8]),
        scs = tonumber(values[o + 9]),
    }
end

function M.parse_serving_cell(text)
    local result = {
        state = nil,
        lte = nil,
        nr5g = nil
    }

    for values in M.parse_response(text, "+QENG") do
        local cell_type = values[1]

        if cell_type == "servingcell" then
            result.state = values[2]

            -- Handle single-line format: "servingcell","NOCONN","LTE","FDD",...
            if values[3] == "LTE" then
                result.lte = parse_lte_fields(values, 4)
            elseif values[3] == "NR5G-NSA" then
                result.nr5g = parse_nr5g_fields(values, 4)
            end

        elseif cell_type == "LTE" then
            result.lte = parse_lte_fields(values, 2)

        elseif cell_type == "NR5G-NSA" then
            result.nr5g = parse_nr5g_fields(values, 2)
        end
    end

    return result
end

--- Parse +QCAINFO response for carrier aggregation
-- @param text AT+QCAINFO response
-- @return Table with pcc (primary) and scc (list of secondary carriers)
--
-- Formats from Quectel RM520N-GL documentation:
--
-- PCC (10 fields):
--   +QCAINFO: "PCC",<earfcn>,<bandwidth>,<band>,<state>,<pcid>,<rsrp>,<rsrq>,<rssi>,<rssnr>
--
-- SCC formats vary by field count:
--   5 fields (NR5G):  "SCC",<arfcn>,<bw_idx>,<band>,<pcid>
--   9 fields:         "SCC",<earfcn>,<bw>,<band>,<state>,<pcid>,<ul_cfg>,<ul_band>,<ul_earfcn>
--   12 fields:        "SCC",<earfcn>,<bw>,<band>,<state>,<pcid>,<ul_cfg>,<ul_band>,<ul_earfcn>,<rsrp>,<rsrq>,<rssnr>
--   13 fields:        "SCC",<earfcn>,<bw>,<band>,<state>,<pcid>,<rsrp>,<rsrq>,<rssi>,<rssnr>,<ul_cfg>,<ul_band>,<ul_earfcn>
--
-- Note: rssnr is NOT the same as SINR from QENG="servingcell". Use serving cell SINR for display.
--
-- Examples:
--   +QCAINFO: "PCC",1350,100,"LTE BAND 3",1,427,-94,-15,-58,-1
--   +QCAINFO: "SCC",275,75,"LTE BAND 1",1,406,-102,-20,-74,-10,0,-,-
--   +QCAINFO: "SCC",648768,10,"NR5G BAND 78",697
function M.parse_qcainfo(text)
    local result = {
        pcc = nil,
        scc = {}
    }

    -- Helper to parse int, returning nil for "-" or empty
    local function parse_int(val)
        if not val or val == "-" or val == "" then return nil end
        return tonumber(val)
    end

    for values in M.parse_response(text, "+QCAINFO") do
        local num_fields = #values
        -- Lua 5.1 compatible: use if instead of goto
        if num_fields >= 5 then
            local role = values[1]  -- "PCC" or "SCC"
            local band_str = values[4] or ""

            -- Parse band string like "LTE BAND 1" or "NR5G BAND 78"
            local rat, band_num = band_str:match("(%w+) BAND (%d+)")
            local is_nr = (rat == "NR5G")

            local carrier = {
                role = role:lower(),
                arfcn = parse_int(values[2]),
                bandwidth_rb = parse_int(values[3]),
                band = tonumber(band_num),
                rat = is_nr and "5g" or "lte",
            }

            if role == "PCC" then
                -- PCC always has 10 fields: role,earfcn,bw,band,state,pci,rsrp,rsrq,rssi,rssnr
                if num_fields >= 10 then
                    carrier.state = parse_int(values[5])
                    carrier.pci = parse_int(values[6])
                    carrier.rsrp = parse_int(values[7])
                    carrier.rsrq = parse_int(values[8])
                    carrier.rssi = parse_int(values[9])
                    carrier.rssnr = parse_int(values[10])
                end
                result.pcc = carrier
            else
                -- SCC format varies by field count
                if num_fields == 5 then
                    -- NR5G SCC: role,arfcn,bw_idx,band,pci
                    carrier.pci = parse_int(values[5])
                elseif num_fields == 9 then
                    -- SCC without signal: role,earfcn,bw,band,state,pci,ul_cfg,ul_band,ul_earfcn
                    carrier.state = parse_int(values[5])
                    carrier.pci = parse_int(values[6])
                    carrier.ul_configured = parse_int(values[7])
                    carrier.ul_band = values[8] ~= "-" and values[8] or nil
                    carrier.ul_earfcn = parse_int(values[9])
                elseif num_fields == 12 then
                    -- SCC with signal (different order): role,earfcn,bw,band,state,pci,ul_cfg,ul_band,ul_earfcn,rsrp,rsrq,rssnr
                    carrier.state = parse_int(values[5])
                    carrier.pci = parse_int(values[6])
                    carrier.ul_configured = parse_int(values[7])
                    carrier.ul_band = values[8] ~= "-" and values[8] or nil
                    carrier.ul_earfcn = parse_int(values[9])
                    carrier.rsrp = parse_int(values[10])
                    carrier.rsrq = parse_int(values[11])
                    carrier.rssnr = parse_int(values[12])
                elseif num_fields >= 13 then
                    -- Full SCC: role,earfcn,bw,band,state,pci,rsrp,rsrq,rssi,rssnr,ul_cfg,ul_band,ul_earfcn
                    carrier.state = parse_int(values[5])
                    carrier.pci = parse_int(values[6])
                    carrier.rsrp = parse_int(values[7])
                    carrier.rsrq = parse_int(values[8])
                    carrier.rssi = parse_int(values[9])
                    carrier.rssnr = parse_int(values[10])
                    carrier.ul_configured = parse_int(values[11])
                    carrier.ul_band = values[12] ~= "-" and values[12] or nil
                    carrier.ul_earfcn = parse_int(values[13])
                end
                table.insert(result.scc, carrier)
            end
        end
    end

    return result
end

--- Parse +QENG="neighbourcell" response
-- @param text AT+QENG="neighbourcell" response
-- @return List of neighbour cells
function M.parse_neighbours(text)
    local neighbours = {}

    for values in M.parse_response(text, "+QENG") do
        local cell_type = values[1]

        if cell_type:match("^neighbourcell") then
            local scope = cell_type:match("neighbourcell (%w+)")  -- "intra" or "inter"
            local rat = values[2]  -- "LTE"

            local neighbour = {
                scope = scope,
                rat = rat:lower(),
                arfcn = tonumber(values[3]),
                pci = tonumber(values[4]),
                rsrq = tonumber(values[5]),
                rsrp = tonumber(values[6]),
                rssi = tonumber(values[7]),
            }

            table.insert(neighbours, neighbour)
        end
    end

    return neighbours
end

--- Parse +QNWPREFCFG response for band configuration
-- @param text AT+QNWPREFCFG response
-- @return setting name, value (string or table of bands)
function M.parse_qnwprefcfg(text)
    for values in M.parse_response(text, "+QNWPREFCFG") do
        local setting = values[1]
        local value = values[2]

        -- Parse band lists like "1:3:7:20"
        if setting:match("band$") and value then
            local bands = {}
            for band in value:gmatch("(%d+)") do
                table.insert(bands, tonumber(band))
            end
            return setting, bands
        end

        return setting, value
    end
    return nil, nil
end

--- Parse +QNWPREFCFG multi-line response for a specific setting
-- Used for ue_capability_band/policy_band which return multiple lines
-- @param text AT+QNWPREFCFG response (may contain multiple +QNWPREFCFG lines)
-- @param target Setting name to find (e.g., "lte_band", "nsa_nr5g_band")
-- @return setting name, value (string or table of bands)
function M.parse_qnwprefcfg_from(text, target)
    for values in M.parse_response(text, "+QNWPREFCFG") do
        local setting = values[1]
        if setting == target then
            local value = values[2]
            if setting:match("band$") and value then
                local bands = {}
                for band in value:gmatch("(%d+)") do
                    table.insert(bands, tonumber(band))
                end
                return setting, bands
            end
            return setting, value
        end
    end
    return nil, nil
end

--- Parse +QNWLOCK response for cell lock status
-- @param text AT+QNWLOCK="common/4g" or "common/5g" response
-- @return Table with type, num_cells, cells list; or nil if not parseable
--
-- 4G format: +QNWLOCK: "common/4g",<num>[,<earfcn>,<pci>,...]
-- 5G format: +QNWLOCK: "common/5g",<num>[,<pci>,<arfcn>,<scs>,<band>,...]
function M.parse_qnwlock(text)
    for values in M.parse_response(text, "+QNWLOCK") do
        local lock_type = values[1]  -- "common/4g" or "common/5g"

        if lock_type == "common/4g" then
            local num_cells = tonumber(values[2]) or 0
            local cells = {}
            -- Pairs of earfcn,pci starting at index 3
            local i = 3
            while i + 1 <= #values do
                table.insert(cells, {
                    earfcn = tonumber(values[i]),
                    pci = tonumber(values[i + 1]),
                })
                i = i + 2
            end
            return { type = "4g", num_cells = num_cells, cells = cells }

        elseif lock_type == "common/5g" then
            local num_cells = tonumber(values[2]) or 0
            local cells = {}
            -- Groups of pci,arfcn,scs,band starting at index 3
            local i = 3
            while i + 3 <= #values do
                table.insert(cells, {
                    pci = tonumber(values[i]),
                    arfcn = tonumber(values[i + 1]),
                    scs = tonumber(values[i + 2]),
                    band = tonumber(values[i + 3]),
                })
                i = i + 4
            end
            return { type = "5g", num_cells = num_cells, cells = cells }
        end
    end
    return nil
end

return M
